cgen error: missing * of optional non-ref structs

```
OS: macos, macOS, 12.6, 21G115
Processor: 8 cpus, 64bit, little endian, Intel(R) Core(TM) i7-8569U CPU @ 2.80GHz
CC version: Apple clang version 14.0.0 (clang-1400.0.29.102)

getwd: /usr/local/v
vmodules: /Users/shove/.vmodules
vroot: /usr/local/v
vexe: /usr/local/v/v
vexe mtime: 2022-10-13 22:53:00
is vroot writable: true
is vmodules writable: true
V full version: V 0.3.1 213a094.f7f0e0b

Git version: git version 2.37.0 (Apple Git-136)
Git vroot status: weekly.2022.40-66-gf7f0e0b5-dirty
.git/config present: true
thirdparty/tcc status: thirdparty-macos-amd64 46662e20
```

**What did you do?**

```v
struct C.stat{}

fn foo() ?C.stat {
	buf := C.stat{}
	return buf
}

fn main() {
	_ := foo() or { C.stat{} }
}
```

**What did you expect to see?**

no error.
 
**What did you see instead?**

C code:

```c
struct _option_struct_stat {
	byte state;
	IError err;
	byte data[sizeof(struct stat) > 1 ? sizeof(struct stat) : 1];        // => missing * of optional non-ref structs
};
```

It should be:

```c
struct _option_struct_stat {
	byte state;
	IError err;
	byte data[sizeof(struct stat*) > 1 ? sizeof(struct stat*) : 1];
};
```

compiler output:

```
==================
/tmp/v_501/a.8877403944744515946.tmp.c:1229:12: error: invalid application of 'sizeof' to an incomplete type 'struct stat'
        byte data[sizeof(struct stat) > 1 ? sizeof(struct stat) : 1];
                  ^     ~~~~~~~~~~~~~
/tmp/v_501/a.8877403944744515946.tmp.c:1229:26: note: forward declaration of 'struct stat'
        byte data[sizeof(struct stat) > 1 ? sizeof(struct stat) : 1];
                                ^
/tmp/v_501/a.8877403944744515946.tmp.c:1229:38: error: invalid application of 'sizeof' to an incomplete type 'struct stat'
        byte data[sizeof(struct stat) > 1 ? sizeof(struct stat) : 1];
                                            ^     ~~~~~~~~~~~~~
/tmp/v_501/a.8877403944744515946.tmp.c:1229:26: note: forward declaration of 'struct stat'
        byte data[sizeof(struct stat) > 1 ? sizeof(struct stat) : 1];
                                ^
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang
```

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -1036,7 +1036,7 @@ fn (g Gen) optional_type_text(styp string, base string) string {
 	} else if base.starts_with('anon_fn') {
 		'void*'
 	} else {
-		base
+		if base.starts_with('struct ') && !base.ends_with('*') { '$base*' } else { base }
 	}
 	ret := 'struct $styp {
 	byte state;
@@ -1053,7 +1053,7 @@ fn (g Gen) result_type_text(styp string, base string) string {
 	} else if base.starts_with('anon_fn') {
 		'void*'
 	} else {
-		base
+		if base.starts_with('struct ') && !base.ends_with('*') { '$base*' } else { base }
 	}
 	ret := 'struct $styp {
 	bool is_error;

--- a/vlib/v/tests/option_test.v
+++ b/vlib/v/tests/option_test.v
@@ -390,6 +390,15 @@ fn test_optional_ref_c_struct_gen() {
 	_ := get_opt_pointer_to_c_struct() or { &C.stat{} }
 }
 
+// For issue #16070: cgen error: missing * of optional non-ref structs
+fn get_opt_to_c_struct() ?C.stat {
+	return none
+}
+
+fn test_optional_c_struct_gen() {
+	_ := get_opt_to_c_struct() or { C.stat{} }
+}
+
 // For issue #16062: checker disallow the return of voidptr(nil) in or block
 struct Bar {}
 

